pkgs/development/libraries/glibc/0001-nss_module.c-try-loading-NSS-modules-from-run-nss-mo.patch

```diff
@@ -0,0 +1,51 @@
+From 65a211ee3604733dceba13062f956256a573f27c Mon Sep 17 00:00:00 2001
+From: Florian Klink <flokli@flokli.de>
+Date: Sun, 19 Sep 2021 13:26:33 +0200
+Subject: [PATCH] nss_module.c: try loading NSS modules from /run/nss-modules
+ as a fallback
+
+On NixOS, glibc only looks for NSS modules in ${glibc.out}/lib, and what
+LD_LIBRARY_PATH is set to.
+
+LD_LIBRARY_PATH is very invasive, so we don't want to set that globally.
+We previously worked this around by running nscd with LD_LIBRARY_PATH
+set, but nscd has some caching issues, and leak of resolution traffic,
+so it's cleaner to have glibc look for NSS modules in an additional
+path, that's provided by NixOS.
+
+On non-NixOS distributions, this shouldn't change behaviour, as the path
+doesn't exist there.
+---
+ nss/nss_module.c | 16 ++++++++++++++++
+ 1 file changed, 16 insertions(+)
+
+diff --git a/nss/nss_module.c b/nss/nss_module.c
+index 6c5f341f..44cfc2a4 100644
+--- a/nss/nss_module.c
++++ b/nss/nss_module.c
+@@ -133,6 +133,22 @@ module_load (struct nss_module *module)
+       return false;
+ 
+     handle = __libc_dlopen (shlib_name);
++
++    /* After loading from the default locations, try loading from
++       /run/nss-modules, to allow loading NixOS-provided NSS modules. */
++    if(handle == NULL)
++      {
++        const char *nix_glibc_nss_path = "/run/nss-modules/";
```

Comment:
Could this not just use:

```suggestion
+        const char *nix_glibc_nss_path = "/usr/lib";
```

Than it would work everywhere not just nixos.

Reply:
Not necessarily, some machines use /lib/x86_64-linux-gnu.

Personally, I'd rather not have this pull in all of the host machine's modules - I'd prefer to be able to pick and choose which ones are picked up by Nix packages.

Reply:
Let's leave the discussion on how the situation on non-NixOS can be improved for a followup PR/issue. This doesn't change behaviour on non-NixOS right now.

Reply:
> Not necessarily, some machines use /lib/x86_64-linux-gnu.
> 
> Personally, I'd rather not have this pull in all of the host machine's modules - I'd prefer to be able to pick and choose which ones are picked up by Nix packages.

How would you do that with this design though?

Reply:
> How would you do that with this design though?

I would symlink the modules I want into a private dir, which is
* made available at /run/nss-modules/lib via the host OS for multi-user non-NixOS, or
* mounted to /run/nss-modules/lib via nix-user-chroot for single-user non-NixOS (happy to contribute this to nix-user-chroot).

I can't think of any situation where someone would be able to create /nix for their nix installation, but be unable to create /run/nss-modules using the same mechanism.

> Let's leave the discussion on how the situation on non-NixOS can be improved for a followup PR/issue. This doesn't change behaviour on non-NixOS right now.

Whether or not this was your intent, this PR is scratching a big itch for non-NixOS. The path you use for nss-module loading is going to be used on non-NixOS (because it's very useful!) and will be depended on. The choice of path deserves careful consideration IMHO.